build-airdrop-list/index.html

```diff
@@ -0,0 +1,204 @@
+      const counterDisplay = document.getElementById('counter-display');
+      const startsAtZeroCheckbox = document.getElementById('starts-at-zero');
+
+      snapshotButton.addEventListener('click', async () => {
+        airDropList = []; // Reset airDropList for each run
+        const collectionAddress = collectionAddressInput.value.trim();
+        const startsAtZero = startsAtZeroCheckbox.checked;
+        if (!collectionAddress) {
+          alert('Please enter a collection address.');
+          return;
+        }
+
+        snapshotButton.disabled = true;
+        snapshotButton.innerText = 'Building Airdrop List...';
+        logBox.value = 'Starting... this can take a while. \nTotal supply: ';
+        let counter = 0;
+
+        try {
+          await buildAirdropList(collectionAddress, startsAtZero, (newCounter) => {
+            counterDisplay.textContent = `Tokens Processed: ${newCounter}`;
+            counter = newCounter;
+          });
+          mergeConsecutiveEntries();
+          downloadCSV(airDropList);
+        } catch (error) {
+          console.log(error);
+          logBox.value += 'Error: ' + error.message + '\n';
+        }
+
+        snapshotButton.disabled = false;
+        snapshotButton.innerText = 'Build Airdrop List';
+      });
+
+      async function buildAirdropList(collectionAddress, startsAtZero, updateCounter) {
+        if (!window.ethereum) {
+          logBox.value += 'Ethereum provider (e.g., MetaMask) not found.\n';
+          return;
+        }
+
+        const provider = new ethers.providers.Web3Provider(window.ethereum);
+        const contract = new ethers.Contract(collectionAddress, [
+          'function ownerOf(uint256 tokenId) external view returns (address)',
+          'function totalSupply() external view returns (uint256)'
+        ], provider);
+
+        const totalSupply = await contract.totalSupply();
+        logBox.value += totalSupply + "\n";
+
+        const batchSize = 100;
+        const startTokenId = startsAtZero ? 0 : 1;
+        const endTokenId = startsAtZero ? totalSupply - 1 : totalSupply;
+        for (let batchStart = startTokenId; batchStart <= endTokenId; batchStart += batchSize) {
+          const batchEnd = Math.min(batchStart + batchSize - 1, endTokenId);
+          const promises = [];
+          for (let tokenId = batchStart; tokenId <= batchEnd; tokenId++) {
+            promises.push((async () => {
+              try {
+                const ownerAddress = await contract.ownerOf(tokenId);
+                const code = await provider.getCode(ownerAddress);
+                const ownerType = code === '0x' ? 'Wallet' : 'Contract';
+                
+                return { tokenId, ownerAddress, ownerType: ownerType, success: true };
+              } catch (error) {
+                return { tokenId, error, success: false };
+              }
+            })());
+          }
+
+          const results = await Promise.allSettled(promises);
```

Comment:
Are these guaranteed to settle in series?

Reply:
Maybe add a sort by token Id if not

Reply:
Actually looks like the return is guaranteed, and since you push to airdrop list afterwards in the forEach it is good:
https://stackoverflow.com/questions/67274188/does-promise-allsettled-preserve-the-original-order-in-the-response-in-all-scena